libs/driver/include/driver/GuiScale.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (C) 2005 - 2023 Settlers Freaks (sf-team at siedler25.org)
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "Point.h"
+#include "helpers/mathFuncs.h"
+#include <type_traits>
+
+/// Represents the scale factors – precomputed from a percentage – to translate between screen and view dimensions
+class GuiScale
+{
+public:
+    constexpr GuiScale(unsigned percent) noexcept
+        : percent_(percent), scaleViewToScreen_(percent_ * .01f), scaleScreenToView_(1.f / scaleViewToScreen_)
+    {}
+
+    constexpr unsigned percent() const noexcept { return percent_; }
+    constexpr float scaleFactor() const noexcept { return scaleViewToScreen_; }
+    constexpr float invScaleFactor() const noexcept { return scaleScreenToView_; }
+
+    /// Translate a scalar from screen to view space
+    template<typename R = float>
+    constexpr R screenToView(float val) const noexcept
+    {
+        if constexpr(std::is_integral_v<R>)
+            return helpers::iround<R>(val * scaleScreenToView_);
+        else
+            return static_cast<R>(val * scaleScreenToView_);
+    }
+
+    /// Translate a scalar from view to screen space
+    template<typename R = float>
+    constexpr R viewToScreen(float val) const noexcept
+    {
+        if constexpr(std::is_integral_v<R>)
+            return helpers::iround<R>(val * scaleViewToScreen_);
+        else
+            return static_cast<R>(val * scaleViewToScreen_);
+    }
+
+    /// Translate a point from screen to view space
+    template<typename R = Position, typename T>
+    constexpr R screenToView(Point<T> pt) const noexcept
```

Comment:
`Point<T>` returns `Point<int>` by default and `float` returns `float` by default. Both functions round when going from floating-point to integer. These choices match the common usage of both overloads.

Reply:
> I don't understand how this relates to this function. `Point<T>` here is whatever is passed in, no defaults here AFAIK.

The return type is defaulted: `typename R = Position` (which is `Point<int>`, right?)

> Using the other overload for the components makes this very clear and even might be faster for float-points as you save a potential copy (you unconditionally construct/cast the input point to `PointF`)

I removed a bunch of template code to do the conversion conditionally after convincing myself that copy elision makes it unnecessary in practice.

> But this is only about being more readable.

More readable where? Here in `GuiScale.h` or at the call site? Because I care more about where the function is used and wanted that to be more compact. I've also only very recently added the `float` overload for clarity.

I don't see how duplicating the `Point` rounding logic here helps anyone. The rounding in `float` is just done for parity. Not that it's actually used anywhere at the moment AFAICR.

Or am I still misunderstanding you?

Reply:
> The return type is defaulted: `typename R = Position` (which is `Point<int>`, right?)

I see, so you meant "`Point<T>` returns `Point<int>`" as in "passing x to **this** function returns y".

> I removed a bunch of template code to do the conversion conditionally after convincing myself that copy elision makes it unnecessary in practice.

Agreed.

> More readable where? Here in `GuiScale.h` or at the call site? 

readable here. You have 2 overloads: one working on a single value and one on a pair of values. From the code here I cannot tell that `foo(pt)` and `Position(foo(pt.x), foo(pt.y))` do the same thing. Which is why I was asking in the first place as the other operations on points are basically implemented that way: Apply the function on each component.   
So it begs the question whether the operation on Points is supposed to be different especially as it duplicates logic: In this case only the choice of `scaleScreenToView_` vs `scaleViewToScreen_`

> I don't see how duplicating the `Point` rounding logic here helps anyone. 

You already duplicated the rounding logic in the single-value function (for IMO good reasons) so why not make use of that avoiding the next person wondering whether there is a difference in the overloads?

At least put the overloads next to each other for discoverability.

Reply:
> At least put the overloads next to each other for discoverability.

No objections here.

> You have 2 overloads: one working on a single value and one on a pair of values. From the code here I cannot tell that `foo(pt)` and `Position(foo(pt.x), foo(pt.y))` do the same thing.

Why though? Multiplying a point (aka. vector) and a scalar is a perfectly well-defined and commonly understood operation.

With casts, this becomes less readable to me:
```cpp
return R(screenToView(static_cast<float>(pt.x)), screenToView(static_cast<float>(pt.y)));
````

And rewriting the scalar overloads is dangerous, as conceptually, a point is more likely to be in a value range that is clearly representable by `float`.
Scalars can represent anything and so I like keeping the cast and the responsibility for it at the call site.

Without the casts:
```cpp
return R(screenToView(pt.x), screenToView(pt.y));
````
This looks fine to me, but requires the dangerous change below:

```cpp
/// Translate a scalar from view to screen space
template<typename R = float, typename T, std::enable_if_t<std::is_arithmetic_v<T>, int> = 0>
constexpr R viewToScreen(T val) const noexcept
{
    if constexpr(std::is_integral_v<R>)
        return helpers::iround<R>(static_cast<float>(val) * scaleViewToScreen_);
    else
        return static_cast<R>(val * scaleViewToScreen_);
}
```

So ultimately, I'd still only group the overloads and keep the rest as-is. :man_shrugging:

Reply:
> Why though? Multiplying a point (aka. vector) and a scalar is a perfectly well-defined and commonly understood operation.

`R(PointF(pt) * scaleScreenToView_) ==  R(screenToView(pt.x), screenToView(pt.y))` isn't immediately obvious due to the rounding when written more accurately: `R(screenToView<R::element_type>(pt.x)...`) 

Anyway your cast argument makes sense although `PointF(pt)` has the exact same issue although yes, it is "less likely". So I guess it is not worth the inconvenience on use.

> So ultimately, I'd still only group the overloads and keep the rest as-is. 🤷‍♂️

Agreed.   
I'd then like to have a test clearly stating `viewToScreen(pt) == Position(viewToScreen<int>(pt.x), viewToScreen<int>(pt.y))` with some random `pt` (via `test::random::randomPoint` or so)